### PR TITLE
Add initial wiki linting

### PR DESCRIPTION
**Pull request by ddbeck**
This is the first step in linting and scraping the wiki. These changes were starting to get a bit big and I had to rework a bit when I realized I needed to consume the unrendered KumaScript macros, so I've decided to break the work of https://github.com/mdn/sprints/issues/2411 into pieces, to make the reviews more manageable. 

If you check out this PR, you can run `npm install` and then `node ./scripts/scraper-ng/ --help` to get started. If you provide an MDN page URL as an argument, you ought to get warnings for KumaScript macros and an explicit failure if you try to lint a page that doesn't have a compat macro, you'll get an error about that.

Things I haven't attempted to deal with yet:

* conversion to Markdown
* writing out Markdown
* making sense of page types and recipes

Feedback I'm especially interested in:

* can you imagine extending this? (e.g., writing more plugins to change the HTML tree or rules to lint the HTML content)? Can you imagine writing some checks for https://github.com/mdn/sprints/issues/2410?
* can you reason about the processing pipeline here? (e.g., is it clear how the script goes from from URL to inspecting the tree?)

But I'm curious what you think, in general, at this early stage.

### CODE REVIEW

File: scripts/scraper-ng/index.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  .describe("q", "Suppress success messages")
+  .alias("q", "quiet")
+  .describe("v", "Expand linter notes, if applicable")
+  .alias("v", "verbose")
+  .boolean(["n", "q", "v"]).argv;
+
+const processor = rehype()
+  .use([require("./plugins/kumascript-macros")])
+  .use([
+    require("./rules/html-has-bcd-table"),
```
Comment (Elchi3) [inline scripts/scraper-ng/index.js]: I think you renamed this at some point but not here.
```suggestion
    require("./rules/html-require-compat-macro"),
```

File: scripts/scraper-ng/index.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
+const rehype = require("rehype");
+const reporter = require("vfile-reporter");
+
+const toVFile = require("./url-to-vfile");
+
+const examplePage = "https://developer.mozilla.org/…/div";
+const examplePath = "https://developer.mozilla.org/…/div";
```
Comment (Elchi3) [inline scripts/scraper-ng/index.js]: examplePath and examplePage are the same. Were you intending to support pages and page trees to scrape?
Also, this isn't a real URL, would it be better to use real examples?

File: scripts/scraper-ng/rules/html-warn-macros.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const rule = require("unified-lint-rule");
+const visit = require("unist-util-visit-parents");
+
+module.exports = rule("html-warn-on-macros", warnOnMacros);
+
+/**
+ * Issue a warning for each macro
+ */
+function warnOnMacros(tree, file) {
```
Comment (Elchi3) [inline scripts/scraper-ng/rules/html-warn-macros.js]: This warns for the compat macro which we require in the other rule. Should we have a list of required / allow listed macros somewhere that both rules could use?

File: scripts/scraper-ng/index.js (hunk 2 above)
Comment (ddbeck) [reply scripts/scraper-ng/index.js]: Yeah, an artifact of my development "process" 😆

I originally thought I'd allow you to supply a bunch of URLs at once. That's… complicated. I'm going to pare this back now, but we'll definitely need a bulk option. More on this in a bit.

Comment (ddbeck) [reply scripts/scraper-ng/index.js]: Oh wait, I misunderstood this. Originally, I was going to let you supply a full URL or a shorthand path (e.g., `https://developer.mozilla.org/en-US/docs/Web/HTML/Element/div` or `/en-US/docs/Web/HTML/Element/div`). Not doing that at all.

Comment (ddbeck) [reply scripts/scraper-ng/index.js]: OK, I've fixed the help text and the code to reflect what's actually happening with one URL at a time. I'll address the page trees question in a comment later.

File: scripts/scraper-ng/rules/html-warn-macros.js (hunk 3 above)
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-warn-macros.js]: Done. Also a nice demonstration of adding preferences to rules. 👍

**Comment by ddbeck** [comment]
I've pushed a few changes addressing your comments, @Elchi3. Thank you!

> The lint-only mode is nice as well, but I would imagine that you want to support page trees as well. A fairly common use case will be to scrape/lint all pages under a given section (say all pages under ttps://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt).

Yes, this is a good point. One thing I haven't coped with yet is how to tell the linter what the rules should be for a given page. It'd be easy enough to map recipes to unified presets (e.g., recipe X has some list of applicable linting rules and other plugins), but I'm not sure how the linter should decide what page type it's looking at. A command line option is easy for one page but, as you say, you'll want to lint big chunks of content.

I've been thinking about a few strategies, but I'd be interested in hearing your thoughts on how you would expect the linter to behave.

**Comment by wbamberg** [comment]
>  I'm not sure how the linter should decide what page type it's looking at

Yes, I had thought a little about that. The only thing I can think is, we will need to use tags, and make sure they are properly set for pages where that matters (that is, pages where we want to apply a more specific recipe than "this is a guide page").

File: scripts/scraper-ng/index.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  .example(`$0 ${examplePage}`, "scrape a page")
+  .example(`$0 -n ${examplePage}`, "dry run (lint)")
+  .demandCommand(1, 1, "A URL is required", "Too many arguments")
+  .describe("n", "Dry run (lint)")
+  .alias("n", "dry-run")
+  .describe("q", "Suppress success messages")
+  .alias("q", "quiet")
+  .describe("v", "Expand linter notes, if applicable")
+  .alias("v", "verbose")
+  .boolean(["n", "q", "v"]).argv;
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/index.js]: Yeah, that's about right. `--quiet` actually works if there's no warnings (e.g., if you comment out the macro warning), but the dry run and verbose options don't have a real effect yet. (The verbose option did, when I was scraping the fully rendered HTML and provided some help text for BCD tables without the right class name.)

File: scripts/scraper-ng/index.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+const processor = rehype()
+  .use([require("./plugins/kumascript-macros")])
+  .use([
+    require("./rules/html-require-compat-macro"),
+    [require("./rules/html-warn-macros"), ["Compat"]]
+  ]);
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/index.js]: Yeah, I can change the way this is done a little bit. I imagine we'll use unified's [preset](https://github.com/unifiedjs/unified#preset) functionality to have different lists of plugins and configuration for different page types. I'll update the current PR to use a preset module though, which should help with readability.

File: scripts/scraper-ng/plugins/kumascript-macros.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+const findAndReplace = require("hast-util-find-and-replace");
+
+/*
+ * Find KumaScript macros in text literal nodes and put them in separate nodes
+ * with data about the macro calls.
+ */
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/plugins/kumascript-macros.js]: Can you give me an example of only matching the first in each node? My initial experimentation with `hast-util-find-and-replace` had me thinking it was finding and replacing all instances within a text node. If it's not matching all of them, that's a bug and I should fix it.

File: scripts/scraper-ng/plugins/kumascript-macros.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+module.exports = kumascriptMacrosPlugin;
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/plugins/kumascript-macros.js]: I'll fix this. I was cribbing from the unified examples and didn't think about this.

File: scripts/scraper-ng/plugins/kumascript-macros.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+const macroRegex = /{{ ?(\w+)\((.*?)\) ?}}/;
+
+function kumascriptMacrosPlugin() {
+  return function transformer(tree) {
+    findAndReplace(tree, macroRegex, match => {
+      const groups = match.match(macroRegex);
+      const macroName = groups[1];
+      const macroParams = groups[2];
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/plugins/kumascript-macros.js]: 👍 Got it. I'll take care of this now.

File: scripts/scraper-ng/rules/html-warn-macros.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+const rule = require("unified-lint-rule");
+const visit = require("unist-util-visit-parents");
+
+module.exports = rule("html-warn-on-macros", warnOnMacros);
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-warn-macros.js]: I didn't get as far with this as I would've liked, but my thinking was that some things might be easier to lint in other forms (e.g., you could lint rehype, then retext, then convert to Markdown). Or we might want to share some plugins between the scraper and the stumptown linter and it'd be easier if we identified what kind of tree the plugin works on.

File: scripts/scraper-ng/url-to-vfile.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    data: { url: url }
+  });
+
+  try {
+    const response = await fetch(toRawMacros(url));
+    if (response.ok) {
+      f.contents = await response.text();
+    } else {
+      const message = f.message(`${response.status} ${response.statusText}`);
+      message.fatal = true;
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/url-to-vfile.js]: Yeah, you're right there's a more tidy way to handle this. I'll fix this to call `f.fail` here (which throws).

File: scripts/scraper-ng/url-to-vfile.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  } catch (err) {
+    f.message(err.message);
+    f.fatal = true;
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/url-to-vfile.js]: Good catch. I'll apply this change.

File: scripts/scraper-ng/url-to-vfile.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    return f;
+  }
+
+  return f;
+}
+
+function toRawMacros(url) {
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/url-to-vfile.js]: I forgot to rename it when I changed what it did. 😳 I'll fix this.

File: scripts/scraper-ng/rules/html-require-compat-macro.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
+const rule = require("unified-lint-rule");
+const visit = require("unist-util-visit");
+
+module.exports = rule("html-require-compat-macro", requireCompatMacro);
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-require-compat-macro.js]: I wanted to demonstrate a simple macro that worked on the tree after I modified it. A real check on compat macros is, as you explained, more complicated. For guide pages we'll allow any number of compat macros and we'll rewrite them as stumptown directives. For JS method pages, for example, we'll want there to be one and only one compat macro under the "Browser compatibility" heading; we'll have specific scraping for that too (into frontmatter and to delete the body text).

But you're right that this check, as written, is general. I'll update it.

File: scripts/scraper-ng/plugins/kumascript-macros.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/plugins/kumascript-macros.js]: Nevermind, I found an example. I'll push a fix for this.

**Comment by ddbeck** [comment]
Pushed a bunch of changes based on Will's comments (thank you, Will). Some notes:

- Moved linting rules config into `preset.js` module (I expect we'll have different presets for different page types, when we get to that level of sophistication)
- `preset.js` takes advantage of `unified-lint-rule`'s severity stuff. Slightly complicates the config, but breaking out into a separate module allows for variable names to spell things out better. I think this is an overall improvement, but I'd like feedback on that especially
- Fixed some macros not being found by the `kumascript-macros` plugin
- Parsed macro parameters
- Started printing macro call parameters for the html-warn-macros rule when the `--verbose` flag is used
- Better error handling on failed page loads (it's still complicated, since I have to handle non-200-300 responses and network errors separately, but it's better)
- Generalized required compat rule to be able to require any macro call
- Moved exports to the end of each file
- Expanded comment on `html-warn-macros`
- Renamed `toRawMacros` to `toRaw`

File: scripts/scraper-ng/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/index.js]: This is much nicer :)

File: scripts/scraper-ng/url-to-vfile.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+  try {
+    const response = await fetch(toRaw(url));
+    if (response.ok) {
+      f.contents = await response.text();
+    } else {
+      f.fail(`${response.status} ${response.statusText}`);
+    }
+  } catch (err) {
+    if (!isVFileMessage(err)) {
```
Comment (wbamberg) [inline scripts/scraper-ng/url-to-vfile.js]: This check seems a bit non-obvious. My first thought was, "in what conditions will `try` throw without there being a fatal error"? But I think what's actually happening is that `fail()` is setting `fatal`. So really the error-handling code is still duplicated, except that one of the copies is hidden inside `fail()`. It would seem clearer to just throw in this case too, then there's only one place it gets handled. But it's a super-minor point, and probaby subjective anyway.

Comment (ddbeck) [reply scripts/scraper-ng/url-to-vfile.js]: I wasn't quite sure what you were getting at here but I think I rubber ducked my way to it (I've left my notes about this below, to show what I was thinking). I guess what should happen here is that I have a wrapper for `fetch` that either re-throws `fetch`'s exceptions or creates a new exception and then have the `vfile` error handling happen in only one place.

I'll push some changes for this shortly.

---

I agree it's a bit hard to follow but I'm not quite sure how to make it more obvious. Here's what I'm trying to capture with the existing code:

* The successful path: `fetch` succeeds at the network level and the response is successful (i.e., the page loads).
* An error response: `fetch` succeeds at the network level but the response is an error (e.g., the server responds with a 404). This corresponds to the `try` block's `else` condition. In the existing code, `f.fail()` throws and is swallowed by the `catch` block.
* Fetch failure: `fetch` itself throws (i.e., it can't make a connection to the server). In the existing code, a new error is created.

In the error cases, I want the `vfile` to have one and only one fatal error attached to it and then return the vfile (i.e., I never want to actually throw from `toVFile` at all).

**Comment by ddbeck** [comment]
@wbamberg Thank you! I think I finally figured out what you meant by the fetch handling. Could you take another look at that?

**Comment by wbamberg** [comment]
Yes, that's exactly what I meant :). Thanks Daniel!

**Comment by ddbeck** [comment]
@Elchi3 Did you want another look at this before merging?

**Comment by ddbeck** [comment]
Thank you both, @Elchi3 and @wbamberg. I think this is a lot better than it was originally. I'll get cracking on the output side of things soon.
